src/kraken.rs

```diff
@@ -0,0 +1,139 @@
+use crate::kraken::btc_dai::{MidMarketRate, Rate};
+use serde::de::Error;
+use serde::Deserialize;
+use std::convert::TryFrom;
+
+/// Fetch Ticker data
+/// More info here: https://www.kraken.com/features/api
+pub async fn get_mid_market_rate() -> anyhow::Result<MidMarketRate> {
+    let rate = reqwest::get("https://api.kraken.com/0/public/Ticker?pair=XBTDAI")
+        .await?
+        .json::<TickerResponse>()
+        .await
+        .map(|response| response.result.xbtdai)?;
+
+    Ok(MidMarketRate::from_kraken(rate))
+}
+
+#[derive(Deserialize)]
+struct TickerResponse {
+    result: Ticker,
+}
+
+#[derive(Deserialize)]
+struct Ticker {
+    #[serde(rename = "XBTDAI")]
+    xbtdai: Rate,
+}
+
+#[derive(Deserialize)]
+struct TickerData {
+    #[serde(rename = "a")]
+    ask: Vec<String>,
+    #[serde(rename = "b")]
+    bid: Vec<String>,
+}
+
+impl TryFrom<TickerData> for Rate {
+    type Error = serde_json::Error;
+
+    fn try_from(value: TickerData) -> Result<Self, Self::Error> {
+        let ask_price = value
+            .ask
+            .first()
+            .ok_or_else(|| serde_json::Error::custom("no ask price"))?;
+        let bid_price = value
+            .bid
+            .first()
+            .ok_or_else(|| serde_json::Error::custom("no bid price"))?;
+
+        Ok(Rate {
```

Comment:
Ok yeah, this is not a `Rate`. You could call it `BidAndAsk`: https://www.investopedia.com/terms/b/bid-and-ask.asp